src/v/cloud_storage/tests/segment_meta_cstore_test.cc

```diff
@@ -1342,3 +1344,551 @@ BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_append_retrieve_edge_case) {
+        auto start_delta = prev.delta_offset_end();
+        auto end_delta = start_delta + p.n_conf;
+        auto ts_base = prev.max_timestamp.value();
+        auto ts_last = ts_base + p.ts_delta;
+        auto t_segm = prev.segment_term() + p.segment_term_delta;
+        auto t_arch = std::max(
+          prev.archiver_term() + p.archiver_term_delta, t_segm);
+
+        if (p.gap_size > 0) {
+            base_offset += p.gap_size;
+        } else if (p.gap_size < 0) {
+            base_offset -= std::min(
+              static_cast<int64_t>(-1 * p.gap_size), base_offset);
+        }
+
+        return make_segment(
+          base_offset,
+          p.n_rec,
+          start_delta,
+          end_delta,
+          ts_base,
+          ts_last,
+          t_arch,
+          t_segm,
+          p.s_size);
+    }
+
+    segment_meta make_segment(
+      int64_t base,
+      int64_t num_records,
+      int64_t start_delta,
+      int64_t end_delta,
+      int64_t base_ts,
+      int64_t last_ts,
+      int64_t archiver_term,
+      int64_t segment_term,
+      uint64_t size) {
+        BOOST_REQUIRE(start_delta <= end_delta);
+        BOOST_REQUIRE(num_records >= 0);
+        BOOST_REQUIRE(size >= 0);
+        BOOST_REQUIRE(archiver_term >= segment_term);
+        segment_meta m{
+          .is_compacted = false,
+          .size_bytes = size,
+          .base_offset = model::offset(base),
+          .committed_offset = model::offset(base + num_records - 1),
+          .base_timestamp = model::timestamp(base_ts),
+          .max_timestamp = model::timestamp(last_ts),
+          .delta_offset = model::offset_delta(start_delta),
+          .archiver_term = model::term_id(archiver_term),
+          .segment_term = model::term_id(segment_term),
+          .delta_offset_end = model::offset_delta(end_delta),
+          .sname_format = segment_name_format::v3,
+          .metadata_size_hint = 0,
+        };
+        return m;
+    }
+
+    bool _add_reuploads;
+    bool _add_reorderings;
+    bool _add_gaps;
+    bool _add_overlaps;
+    bool _add_empty;
+    std::map<model::offset, segment_meta> _segments;
+    /// Used to generate out of order updates
+    std::vector<segment_meta> _ooo_write;
+    int _num_truncations{0};
+};
+
+void test_segment_meta_cstore_invariants_randomized(
+  const int n_iter,
+  bool allow_reuploads,
+  bool allow_reorderings,
+  bool allow_gaps,
+  bool allow_overlaps,
+  bool allow_empty) {
+    segment_meta_random_walk walk(
+      allow_reuploads,
+      allow_reorderings,
+      allow_gaps,
+      allow_overlaps,
+      allow_empty);
+
+    segment_meta_cstore cstore;
+    vlog(test.info, "Running randomized invariants test");
+    vlog(
+      test.info,
+      "n_iter = {}, allow_reuploads = {}, allow_gaps = {}, allow_overlaps = "
+      "{}, allow_empty = {}",
+      n_iter,
+      allow_reuploads,
+      allow_gaps,
+      allow_overlaps,
+      allow_empty);
+
+    auto cond =
```

Comment:
nit: it'd have been great if this were a gtest, then some of this could be done via `ASSERT_EQ/NE(rhs, lhs) << "msg"`

Reply:
Also, it feels like `cond` and `cond1` should be merged? Maybe I'm missing something but are they meaningfully different enough to separate them?